phabricatoremails/render/events/common.py

```diff
@@ -0,0 +1,79 @@
+from dataclasses import dataclass
+from datetime import timezone, timedelta
+from enum import Enum
+from math import floor, ceil
+from typing import Optional, Dict, List
+
+"""Describes data structures that are used by both secure and public revision events.
+
+Additional information about this module can be found in __init__.py
+"""
+
+
+@dataclass
+class Recipient:
+    email: str
+    timezone: timezone
+    is_actor: bool
+
+    @staticmethod
+    def parse(recipient: Dict):
+        offset_seconds = recipient["timezoneOffset"]  # e.g.: -25200
+        if offset_seconds > 0:
+            hours = floor(offset_seconds / 3600)
+            minutes = floor((offset_seconds - hours * 3600) / 60)
+        else:
+            hours = ceil(offset_seconds / 3600)
+            minutes = ceil((offset_seconds - hours * 3600) / 60)
+        return Recipient(
+            recipient["email"],
+            timezone(timedelta(hours=hours, minutes=minutes)),
+            recipient["isActor"],
+        )
+
+    @staticmethod
+    def parse_optional(recipient: Optional[Dict]):
```

Comment:
I'm not sure how it's hacky, it's here for ergonomic purposes, and it's named accordingly.
Optional recipients are pretty common (e.g.: revision authors who haven't enabled Mozilla emails), and repeating the `parse(...) if "recipient" in obj else None` pattern was pretty common and worth factoring away.

If there's a better name or technique here for representing a "parse this `Optional`" function, let me know.

Reply:
I see your point here, but is there any reason not to include this in `parse` if it's such a common case? (Or, perhaps `parse` should have a `required` argument that raises an exception when set to `True` (default behaviour), or returns `None` if set to `False` and if it receives a `None`.

Reply:
> The comment you linked to is the same as the first comment on this line -- i.e. moving the check to where you are using it.

Github can be goofy at opening links on the same PR. Try opening [the link](https://github.com/mozilla-conduit/phabricator-emails/pull/1#discussion_r404886661) in a new tab, instead?

![Screenshot from 2020-04-16 10-54-00](https://user-images.githubusercontent.com/7784737/79489739-9badb880-7fd0-11ea-9850-bc2b47a43e1e.png)


> You are maing assumptions as to if someone who might be using your code is using an IDE or what IDE they are using. For example, I use Vim and no IDE.

I agree that we shouldn't implement IDE-specific assumptions in our code, like Vim modelines or IntelliJ lint suppressions. However, there's a **cost** in not embracing tool-agnostic techniques. And, in the future, you might find an advanced Python plugin for Vim that provides type-based intellisense, or we could be using `mypy`.

> This is true, but the failure will be a KeyError with regards to timezoneOffset. Less user friendly than saying "a required argument was ommitted".

If this was a public interface of a library, then I agree that we should be defensively validating arguments. However, since this is internal, I don't think that adding logic to improve the type of exception is worth it. There's other cases throughout this codebase in which we receive a parameter that we assume is set, and will spontaneously combust if `None` is provided instead. I don't think that we should defensively check-and-throw-exception in those locations, too.